### PR TITLE
Support gRPC config for agent-service plugin

### PR DESCRIPTION
**Pull request by honnix**
# TL;DR
Support customized gRPC config for agent-service plugin.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [x] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description

Support configuring gRPC endpoint with:

* insecure
* default service config
* timeouts

Details can be found in https://github.com/flyteorg/flyte/issues/3823

**Note that this is a breaking change!** I'm not sure how mature the `agent-service` is and how many users are using it, but if backward compatibility is important in this case, I can go an extra mile to achieve that.

## Tracking Issue
Closes https://github.com/flyteorg/flyte/issues/3823

## Follow-up issue
_NA_

### CODE REVIEW

File: go/tasks/plugins/webapi/agent/plugin.go [hunk 1]
```diff
@@ -144,28 +157,44 @@ func (p Plugin) Status(ctx context.Context, taskCtx webapi.StatusContext) (phase
...
+		pool, err := x509.SystemCertPool()
+		if err != nil {
+			return nil, err
+		}
+
+		creds := credentials.NewClientTLSFromCert(pool, "")
+		opts = append(opts, grpc.WithTransportCredentials(creds))
+	}
+
+	if endpoint.DefaultServiceConfig != "" {
```
Comment (hamersaw) [inline go/tasks/plugins/webapi/agent/plugin.go:188]: I think `len(endpoint.DefaultServiceConfig) > 0` is the idiomatic golang way to do this.

File: go/tasks/plugins/webapi/agent/config.go [hunk 2]
```diff
@@ -10,6 +10,8 @@ import (
 )
 
 var (
+	defaultTimeout = config.Duration{Duration: 10 * time.Second}
```
Comment (hamersaw) [inline go/tasks/plugins/webapi/agent/config.go:13]: do we want to make this configurable?

**Review by hamersaw — COMMENTED** [review]
@honnix can you elaborate on what is not backwards compatible?

File: go/tasks/plugins/webapi/agent/config.go (hunk 2 above)
Comment (honnix) [reply go/tasks/plugins/webapi/agent/config.go:13]: Good question. I thought about how to deal with the default here. My reasoning is that if user bothers configuring this, they might just configure for each operation instead because it doesn't save much effort. That being said, I don't have a strong opinion and I can make it configurable for sure.

File: go/tasks/plugins/webapi/agent/plugin.go (hunk 1 above)
Comment (honnix) [reply go/tasks/plugins/webapi/agent/plugin.go:188]: Sure. I just followed existing code I found.

**Comment by honnix** [comment]
> @honnix can you elaborate on what is not backwards compatible?

Both `EndpointForTaskTypes` and `DefaultGrpcEndpoint` are now completely different types so if a user has configured either of those, upgrading to this version will crash their application I guess.

File: go/tasks/plugins/webapi/agent/config.go [hunk 3]
```diff
@@ -78,8 +77,11 @@ type GrpcEndpoint struct {
 	// DefaultServiceConfig sets default gRPC service config; check https://github.com/grpc/grpc/blob/master/doc/service_config.md for more details
 	DefaultServiceConfig string `json:"defaultServiceConfig"`
 
-	// Timeouts defines various RPC timeout values for different plugin operations: CreateTask, GetTask, DeleteTask; if not configured, defaults to 10s
+	// Timeouts defines various RPC timeout values for different plugin operations: CreateTask, GetTask, DeleteTask; if not configured, defaults to DefaultTimeout
 	Timeouts map[string]config.Duration `json:"timeouts"`
+
+	// DefaultTimeout gives the default RPC timeout if a more specific one is not defined in Timeouts
+	DefaultTimeout config.Duration `json:"defaultTimeout"`
```
Comment (honnix) [inline go/tasks/plugins/webapi/agent/config.go:84]: Note that this is now a mandatory config item, otherwise it defaults to `0` I guess, which would fail the RPC immediately.

Comment (honnix) [reply go/tasks/plugins/webapi/agent/config.go:84]: This doesn't seem to be nice so I addressed it in 2a706c3e85d7eac4ff4d5214f825f7813cbdb60a

**Comment by hamersaw** [comment]
> upgrading to this version will crash their application I guess

@pingsutw we released flyte agents as an experimental feature right? so no guarantees of backwards compatibility. IMO fixing this grpc configuration issue warrants a backwards incompatible merge. thoughts?

**Comment by eapolinario** [comment]
> > upgrading to this version will crash their application I guess
> 
> @pingsutw we released flyte agents as an experimental feature right? so no guarantees of backwards compatibility. IMO fixing this grpc configuration issue warrants a backwards incompatible merge. thoughts?

Agreed. This feature is tagged as experimental, so let's use this time to break backwards compatibility if needed (which seems like this particular instance is one of such times).

**Comment by hamersaw** [comment]
@honnix thinking about this it will be unable to cover the scenario where there are two flyte agent deployments where each satisfies multiple task types right? I'm thinking something like:
```
grpcConnections:
    - endpoint: x.com:8088
      taskTypes:
      - foo
      - bar
    - endpoint: y.com:8088
      taskTypes:
      - baz
      - bat
```
the current approach would create a separate gRPC connection for each task type right? whereas it would probably be better to reuse a single connection for each endpoint. Thoughts?

**Comment by honnix** [comment]
I don't think so. The task type here is only for routing and there should be only one grpc connection per each endpoint. This PR shouldn't change how that part works I think.

**Comment by hamersaw** [comment]
> I don't think so. The task type here is only for routing and there should be only one grpc connection per each endpoint. This PR shouldn't change how that part works I think.

So the [EndpointForTaskTypes map](https://github.com/flyteorg/flyteplugins/blob/3d2525a9f9373a220219fb2d1c07f0b38101b23b/go/tasks/plugins/webapi/agent/config.go#L60) is keyed on the task type, but the [connectionCache](https://github.com/flyteorg/flyteplugins/blob/3d2525a9f9373a220219fb2d1c07f0b38101b23b/go/tasks/plugins/webapi/agent/plugin.go#L30) is keyed on the endpoint. So you can't have two endpoints configured differently for different task types. For example, differing timeouts will result in unintended behavior:
```
endpointForTaskTypes:
  foo: 
    endpoint: localhost:8088
    timeout: 20
  bar: 
    endpoint: localhost:8088
    timeout: 40
```
IIUC the above would only create a single gRPC connection in the background (with timeout of whatever agent was called first) and reuse it for calls to both task types. I don't think this is a huge issue, but I don't like the unintended behavior for future added gRPC params that might be more important.

**Comment by honnix** [comment]
@hamersaw I did some experiment in https://github.com/flyteorg/flyteplugins/pull/368/commits/d67a01563a504ab1cf33940056e1ca43b3ed3c38 and now the config is like:

```yaml
endpointForTaskTypes:
  foo: agent
  bar: bar_agent
  foo_bar: agent

grpcEndpoints:
  agent:
    endpoint: localhost:8088
    timeouts:
      CreateTask: 1s
  bar_agent:
    endpoint: localhost:8088
    timeouts:
      CreateTask: 2s
```

And the cache is keyed on the whole endpoint config (as a pointer).

**Comment by honnix** [comment]
> Does it make sense to update the naming here? Ex `grpcEndpoints` -> `agentEndpoints` and `endpointForTaskTypes` -> `agentsForTaskTypes` or `taskTypeAgents`?

SGTM but maybe it's better double check with the original author of the config section.

> Also, just want to make sure this doesn't introduce additional confusion. We still need to add all supported plugins to the `supportedTaskTypes` configuration right?

Yes that is correct. This PR does not make any change to the semantics of `supportedTaskTypes`.

**Comment by honnix** [comment]
> Does it make sense to update the naming here? Ex `grpcEndpoints` -> `agentEndpoints` and `endpointForTaskTypes` -> `agentsForTaskTypes` or `taskTypeAgents`?
> 
> Also, just want to make sure this doesn't introduce additional confusion. We still need to add all supported plugins to the `supportedTaskTypes` configuration right?

@hamersaw I did some experiments in the latest commit. PTAL, thanks!

**Review by hamersaw — APPROVED** [review]

**Review by pingsutw — APPROVED** [review]
@honnix Thank you so much

**Comment by honnix** [comment]
Thank you all for reviewing and providing great comments! No worries about nits at all. If we are gonna ship a breaking change, it's better we do it right.

**Comment by pingsutw** [comment]
@honnix btw, would you mind helping us update the [doc](https://github.com/flyteorg/flytesnacks/blob/5c5380d28cec7ae57d646b7e0cd6bce195c1df5a/examples/extend_flyte/extend_flyte/agent_service.py#L171-L186) here.

**Comment by honnix** [comment]
> @honnix btw, would you mind helping us update the [doc](https://github.com/flyteorg/flytesnacks/blob/5c5380d28cec7ae57d646b7e0cd6bce195c1df5a/examples/extend_flyte/extend_flyte/agent_service.py#L171-L186) here.

Sure. I did not know this. I can do it tomorrow.
